README.rst
.. image:: https://secure.travis-ci.org/collective/collective.edtf_behavior.png?branch=master
    :target: http://travis-ci.org/collective/collective.edtf_behavior

.. image:: https://coveralls.io/repos/github/collective/collective.edtf_behavior/badge.svg?branch=master
    :target: https://coveralls.io/github/collective/collective.edtf_behavior?branch=master
    :alt: Coveralls

.. image:: https://img.shields.io/pypi/l/collective.edtf_behavior.svg
    :target: https://pypi.python.org/pypi/collective.edtf_behavior/
    :alt: License

.. image:: https://badges.gitter.im/collective/collective.edtf_behavior.svg
   :alt: Join the chat at https://gitter.im/collective/collective.edtf_behavior
   :target: https://gitter.im/collective/collective.edtf_behavior?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge


========================
collective.edtf_behavior
========================

A Plone behavior which provides a `EDTF <http://www.loc.gov/standards/datetime/edtf.html>`_  Date field and some indexes to search and sort on complex and historic dates.


Features
--------

- Allows unspecific dates or date intervals: "1989-11" or "1989-11~" or "1989-11-01/1989-11-30"
- Seasons: 2001-21  >> Spring, 2001
- old dates like 03.08.1492
- Search indexes: date_earliest and date_latest
- Sort indexes: date_sort_ascending and date_sort_descending


Installation
------------

Install collective.edtf_behavior by adding it to your buildout::

    [buildout]

    ...

    eggs =
        collective.edtf_behavior


and then running ``bin/buildout``

After you activated the add-on in the Plone Add-ons section, you can Enable this behavior on any Dexterity based content type.

Querying
--------

.. code-block:: python

        from plone import api

find any document which earliest date is 06.02.1920 or later.

.. code-block:: python

        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': datetime.date(1920, 2, 6),
                'range': 'min',
            },
        )

find any document which latest date is 11.11.1711 or earlier.

.. code-block:: python

        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': datetime.date(1711, 11, 11),
                'range': 'max',
            },
        )

This package provides a `DateRangeInRangeIndex <https://pypi.org/project/Products.DateRangeInRangeIndex/>`_  named ``edtf_start_end_range``.

find all documents that fall at least with one start/end date of there interval into the range:

.. code-block:: python

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1711, 11, 11),
                'end':   datetime.date(1920, 2, 6),
            },
        )

For more examples how to query the indexes, have a look at the `tests <https://github.com/collective/collective.edtf_behavior/tree/master/src/collective/edtf_behavior/tests>`_ .


Extending it
------------

One could improve indexing by using a DateRangeIndex, like the effectiveRange index in Plone if needed.

.. code-block:: xml

    <!-- Example of a DateRangeIndex like the effectiveRange index in Plone -->
    <index name="edtf_date_range" meta_type="DateRangeIndex"
      since_field="date_earliest" until_field="date_latest" />


Contribute
----------

- Issue Tracker: https://github.com/collective/collective.edtf_behavior/issues
- Source Code: https://github.com/collective/collective.edtf_behavior


Support
-------

If you are having issues, please let us know.
You can reach us on Gitter.


License
-------

The project is licensed under the GPLv2.


CHANGES.rst
Changelog
=========


2.0.0 (2023-10-18)
------------------

- using edtf2 package which implements the newest edtf specification (2019-02-04) [muellers]
- adjusting indexe to convert infinity results to datetime.min and datetime.max; in general the date_latest and date_earliest indexe are based on pythons datetime package where we have the 1 AD - 9999 AD limitation, so they are not useable for dates outside this limitation [muellers]

1.0.4 (2023-10-18)
------------------

- fixed wrong python_requires statement in setup.py [muellers]


1.0.3 (2019-08-26)
------------------

- Faxed test and test setup, small cleanup
  [MrTango]


1.0.2 (2019-03-22)
------------------

- Add edtf_parseable validator (constraint) to edtf_behavior
  [MrTango]


1.0.1 (2019-03-18)
------------------

- Add absolute_import statement to all files, to improve Py3 support, fix README syntax
  [MrTango]


1.0 (2019-03-18)
----------------

- Add edtf_start_end_range index and document querying in the readme.
  [MrTango]


1.0a1 (2019-03-14)
------------------

- Initial release, with EDTF behavior and 4 main indexes.
  [MrTango]


src/collective/edtf_behavior/tests/test_behavior_edtf_date.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from collective.edtf_behavior.behaviors.edtf_date import edtf_parseable
from collective.edtf_behavior.behaviors.edtf_date import IEDTFDateMarker
from collective.edtf_behavior.testing import COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING  # noqa
from plone import api
from plone.app.testing import setRoles
from plone.app.testing import TEST_USER_ID
from plone.behavior.interfaces import IBehavior
from zope.component import getUtility
from zope.interface import Invalid

import datetime
import unittest


class EDTFDateIntegrationTest(unittest.TestCase):

    layer = COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING

    def setUp(self):
        """Custom shared utility setup for tests."""
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])
        self.document = api.content.create(
            container=self.portal,
            type='Document',
            id='doc1',
        )

    def test_get_behavior_edtf_date(self):
        behavior = getUtility(IBehavior, 'collective.edtf_behavior.edtf_date')
        self.assertEqual(
            behavior.marker,
            IEDTFDateMarker,
        )

    def test_behavior_edtf_date_earliest_latest(self):
        behavior = getUtility(IBehavior, 'collective.edtf_behavior.edtf_date')
        doc_adapted = behavior.factory(self.document)

        doc_adapted.edtf_date = '1711-11-11'
        self.assertEqual(
            doc_adapted.date_earliest,
            datetime.date(1711, 11, 11),
        )
        self.assertEqual(
            doc_adapted.date_latest,
            datetime.date(1711, 11, 11),
        )

        doc_adapted.edtf_date = '1920-02-03/1933-04-30'
        self.assertEqual(
            doc_adapted.date_earliest,
            datetime.date(1920, 2, 3),
        )
        self.assertEqual(
            doc_adapted.date_latest,
            datetime.date(1933, 4, 30),
        )

        doc_adapted.edtf_date = '1811-12-11/1860-03-13'
        self.assertEqual(
            doc_adapted.date_earliest,
            datetime.date(1811, 12, 11),
        )
        self.assertEqual(
            doc_adapted.date_latest,
            datetime.date(1860, 3, 13),
        )

    def test_behavior_edtf_date_sort_asc_desc(self):
        behavior = getUtility(IBehavior, 'collective.edtf_behavior.edtf_date')
        doc_adapted = behavior.factory(self.document)

        doc_adapted.edtf_date = '1711-11-11'
        self.assertEqual(
            doc_adapted.date_sort_ascending,
            datetime.date(1711, 11, 11),
        )
        self.assertEqual(
            doc_adapted.date_sort_descending,
            datetime.date(1711, 11, 11),
        )

        doc_adapted.edtf_date = '1920-03~'
        self.assertEqual(
            doc_adapted.date_sort_ascending,
            datetime.date(1920, 3, 1),
        )
        self.assertEqual(
            doc_adapted.date_sort_descending,
            datetime.date(1920, 3, 31),
        )

        doc_adapted.edtf_date = '1811-12-11/1860-03-13'
        self.assertEqual(
            doc_adapted.date_sort_ascending,
            datetime.date(1811, 12, 11),
        )
        self.assertEqual(
            doc_adapted.date_sort_descending,
            datetime.date(1860, 3, 13),
        )

    def test_behavior_edtf_validator(self):
        with self.assertRaises(Invalid):
            res = edtf_parseable('12.12.2018')
            self.assertFalse(res)

        res = edtf_parseable('2018-12-12')
        self.assertTrue(res)

        res = edtf_parseable('2018-12~')
        self.assertTrue(res)


src/collective/edtf_behavior/tests/test_behavior_e_d_t_f_date.py
# -*- coding: utf-8 -*-
from collective.edtf_behavior.behaviors.e_d_t_f_date import IEDTFDateMarker
from collective.edtf_behavior.testing import COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING  # noqa
from plone.app.testing import setRoles
from plone.app.testing import TEST_USER_ID
from plone.behavior.interfaces import IBehavior
from zope.component import getUtility

import unittest


class EDTFDateIntegrationTest(unittest.TestCase):

    layer = COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING

    def setUp(self):
        """Custom shared utility setup for tests."""
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])

    def test_behavior_e_d_t_f_date(self):
        behavior = getUtility(IBehavior, 'collective.edtf_behavior.e_d_t_f_date')
        self.assertEqual(
            behavior.marker,
            IEDTFDateMarker,
        )


src/collective/edtf_behavior/tests/test_edtf_search_on_indexes.py
# -*- coding: utf-8 -*-
from collective.edtf_behavior.behaviors.edtf_date import IEDTFDate
from collective.edtf_behavior.behaviors.edtf_date import IEDTFDateMarker
from collective.edtf_behavior.testing import COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING  # noqa
from plone import api
from plone.app.testing import setRoles
from plone.app.testing import TEST_USER_ID
from plone.dexterity.interfaces import IDexterityFTI
from plone.dexterity.schema import SchemaInvalidatedEvent
from zope.component import queryUtility
from zope.event import notify

import datetime
import unittest


class EDTFDateSearchIntegrationTest(unittest.TestCase):

    layer = COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING

    def setUp(self):
        """Custom shared utility setup for tests."""
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])

        # enable IEDTFDate behavior on Document
        fti = queryUtility(IDexterityFTI, name='Document')
        behaviors = list(fti.behaviors)
        behaviors.append(IEDTFDate.__identifier__)
        fti.behaviors = tuple(behaviors)
        notify(SchemaInvalidatedEvent('Document'))

        self.document1 = api.content.create(
            container=self.portal,
            type='Document',
            id='doc1',
            edtf_date='1711-11-11',
        )
        self.document2 = api.content.create(
            container=self.portal,
            type='Document',
            id='doc2',
            edtf_date='1920-02-03/1933-04-30',
        )
        self.document3 = api.content.create(
            container=self.portal,
            type='Document',
            id='doc3',
            edtf_date='1920-02-06/1933-04-28',
        )

    def test_edtf_date_behavior_is_enabled(self):

        self.assertTrue(
            IEDTFDateMarker.providedBy(self.document1),
        )

    def test_edtf_date_search_on_latest(self):
        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': datetime.date(1933, 4, 30),
                'range': 'max',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': datetime.date(1933, 4, 29),
                'range': 'max',
            },
        )
        self.assertEqual(
            len(results),
            2,
        )

        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': datetime.date(1711, 11, 11),
                'range': 'max',
            },
        )
        self.assertEqual(
            len(results),
            1,
        )

        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': datetime.date(1711, 11, 11),
                'range': 'min',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            date_latest={
                'query': (
                    datetime.date(1711, 11, 11),
                    datetime.date(1933, 4, 30),
                ),
                'range': 'min:max',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

    def test_edtf_date_search_on_earliest(self):
        # 1920-02-03 / 1920-02-06 / 1711-11-11
        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': datetime.date(1711, 11, 11),
                'range': 'min',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': datetime.date(1920, 2, 3),
                'range': 'min',
            },
        )
        self.assertEqual(
            len(results),
            2,
        )

        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': datetime.date(1920, 2, 6),
                'range': 'min',
            },
        )
        self.assertEqual(
            len(results),
            1,
        )

        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': datetime.date(1920, 2, 6),
                'range': 'max',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            date_earliest={
                'query': (
                    datetime.date(1711, 11, 11),
                    datetime.date(1920, 2, 6),
                ),
                'range': 'min:max',
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

    def test_edtf_date_search_on_edtf_date_range(self):
        # 1920-02-03 / 1920-02-06 / 1711-11-11

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1711, 11, 11),
                'end':   datetime.date(1920, 2, 6),
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1711, 11, 11),
                'end':   datetime.date(1920, 2, 4),
            },
        )
        self.assertEqual(
            len(results),
            2,
        )

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1711, 11, 11),
                'end':   datetime.date(1933, 4, 29),
            },
        )
        self.assertEqual(
            len(results),
            3,
        )

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1933, 4, 1),
                'end':   datetime.date(1933, 5, 1),
            },
        )
        self.assertEqual(
            len(results),
            2,
        )

        results = api.content.find(
            portal_type='Document',
            edtf_start_end_range={
                'start': datetime.date(1933, 4, 29),
                'end':   datetime.date(1933, 5, 1),
            },
        )
        self.assertEqual(
            len(results),
            1,
        )

    def test_edtf_date_sort_descending(self):
        results = api.content.find(
            portal_type='Document',
            sort_on='date_sort_descending',
            sort_order='descending',
        )
        result_list = [res.getObject().edtf_date for res in results]
        # most recent first
        self.assertEqual(
            result_list,
            ['1920-02-03/1933-04-30', '1920-02-06/1933-04-28', '1711-11-11'],
        )

    def test_edtf_date_sort_ascending(self):
        results = api.content.find(
            portal_type='Document',
            sort_on='date_sort_ascending',
            sort_order='ascending',
        )
        # most recent last
        result_list = [res.getObject().edtf_date for res in results]
        self.assertEqual(
            result_list,
            ['1711-11-11', '1920-02-03/1933-04-30', '1920-02-06/1933-04-28'],
        )


src/collective/edtf_behavior/testing.py
# -*- coding: utf-8 -*-
from plone.app.contenttypes.testing import PLONE_APP_CONTENTTYPES_FIXTURE
from plone.app.robotframework.testing import REMOTE_LIBRARY_BUNDLE_FIXTURE
from plone.app.testing import applyProfile
from plone.app.testing import FunctionalTesting
from plone.app.testing import IntegrationTesting
from plone.app.testing import PloneSandboxLayer
from plone.testing import z2

import collective.edtf_behavior


class CollectiveEdtfBehaviorLayer(PloneSandboxLayer):

    defaultBases = (PLONE_APP_CONTENTTYPES_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        # Load any other ZCML that is required for your tests.
        # The z3c.autoinclude feature is disabled in the Plone fixture base
        # layer.
        import plone.restapi
        self.loadZCML(package=plone.restapi)
        self.loadZCML(package=collective.edtf_behavior)

    def setUpPloneSite(self, portal):
        applyProfile(portal, 'collective.edtf_behavior:default')


COLLECTIVE_EDTF_BEHAVIOR_FIXTURE = CollectiveEdtfBehaviorLayer()


COLLECTIVE_EDTF_BEHAVIOR_INTEGRATION_TESTING = IntegrationTesting(
    bases=(COLLECTIVE_EDTF_BEHAVIOR_FIXTURE,),
    name='CollectiveEdtfBehaviorLayer:IntegrationTesting',
)


COLLECTIVE_EDTF_BEHAVIOR_FUNCTIONAL_TESTING = FunctionalTesting(
    bases=(COLLECTIVE_EDTF_BEHAVIOR_FIXTURE,),
    name='CollectiveEdtfBehaviorLayer:FunctionalTesting',
)


COLLECTIVE_EDTF_BEHAVIOR_ACCEPTANCE_TESTING = FunctionalTesting(
    bases=(
        COLLECTIVE_EDTF_BEHAVIOR_FIXTURE,
        REMOTE_LIBRARY_BUNDLE_FIXTURE,
        z2.ZSERVER_FIXTURE,
    ),
    name='CollectiveEdtfBehaviorLayer:AcceptanceTesting',
)


src/collective/edtf_behavior/upgrades.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from plone.app.upgrade.utils import loadMigrationProfile
from plone import api
import transaction
from .behaviors.edtf_date import IEDTFDate, IEDTFDateMarker
import edtf2
from edtf2.parser.edtf_exceptions import EDTFParseException
import logging

logger = logging.getLogger('collective.edtf_bahviour.upgrades')


def reload_gs_profile(context):
    loadMigrationProfile(
        context,
        'profile-collective.edtf_behavior:default',
    )


def to_1002(context):
    # for each existing edtf_date: check if its parsable; if not, try to convert it
    count_total = 0
    count_values = 0
    count_nochanges = 0
    count_parser_errors = 0
    count_other_errors = 0
    count_converts = 0

    # find all of our etdf values
    edtfs = api.content.find(object_provides=[IEDTFDateMarker, IEDTFDate])
    logger.info("Starting to check edtf_date values of {0} items".format(len(edtfs)))
    for brain in edtfs:
        obj = brain.getObject()
        adapted_obj = IEDTFDate(obj)
        count_total = count_total + 1
        if adapted_obj.edtf_date:
            value = adapted_obj.edtf_date
            count_values = count_values + 1
            try:
                # if its parsable nothing has to be done
                edtf2.parse_edtf(value)
                count_nochanges = count_nochanges + 1
            except EDTFParseException:
                # if there is ParseException, try to convert value to new specification and try to parse again
                new_expression = edtf2.old_specs_to_new_specs_expression(value)
                try:
                    edtf2.parse_edtf(new_expression)
                    count_converts = count_converts + 1
                    adapted_obj.edtf_date = str(new_expression)
                    obj.reindexObject()

                except EDTFParseException:
                    count_parser_errors = count_parser_errors + 1
                except Exception:
                    count_other_errors = count_other_errors + 1
            except Exception:
                count_other_errors = count_other_errors + 1
        if count_total % 100 == 0:
            logger.info("processed {0} from {1} items ...".format(count_total, len(edtfs)))

    logger.info("total {0} setted edtf_date values from {1} items".format(count_values, len(edtfs)))
    logger.info("Values with no changes: {0}; values succesfully converted: {1}; values with parse errors: {2}; values with some other errors: {2};".format(count_nochanges, count_converts, count_parser_errors, count_other_errors))
    transaction.commit()


src/collective/edtf_behavior/indexers/edtf.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from collective.edtf_behavior.behaviors.edtf_date import IEDTFDate
from collective.edtf_behavior.behaviors.edtf_date import IEDTFDateMarker
from plone.dexterity.interfaces import IDexterityContent
from plone.indexer import indexer


@indexer(IDexterityContent)
def dummy(obj):
    """Dummy to prevent indexing child objects"""
    raise AttributeError('This field should not indexed here!')


@indexer(IEDTFDateMarker)
def date_latest_idx(obj):
    adapted_obj = IEDTFDate(obj)
    return adapted_obj.date_latest


@indexer(IEDTFDateMarker)
def date_earliest_idx(obj):
    adapted_obj = IEDTFDate(obj)
    test = adapted_obj.date_earliest
    return test


@indexer(IEDTFDateMarker)
def date_sort_ascending_idx(obj):
    adapted_obj = IEDTFDate(obj)
    return adapted_obj.date_sort_ascending


@indexer(IEDTFDateMarker)
def date_sort_descending_idx(obj):
    adapted_obj = IEDTFDate(obj)
    return adapted_obj.date_sort_descending


src/collective/edtf_behavior/locales/update.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import
import os
import pkg_resources
import subprocess


domain = 'collective.edtf_behavior'
os.chdir(pkg_resources.resource_filename(domain, ''))
os.chdir('../../../')
target_path = 'src/collective/edtf_behavior/'
locale_path = target_path + 'locales/'
i18ndude = './bin/i18ndude'


def locale_folder_setup():
    os.chdir(locale_path)
    languages = [d for d in os.listdir('.') if os.path.isdir(d)]
    for lang in languages:
        folder = os.listdir(lang)
        if 'LC_MESSAGES' in folder:
            continue
        else:
            lc_messages_path = lang + '/LC_MESSAGES/'
            os.mkdir(lc_messages_path)
            cmd = 'msginit --locale={0} --input={1}.pot --output={2}/LC_MESSAGES/{3}.po'.format(   # NOQA: E501
                        lang,
                        domain,
                        lang,
                        domain,
                    )
            subprocess.call(
                cmd,
                shell=True,
            )

    os.chdir('../../../../')


def _rebuild():
    cmd = '{0} rebuild-pot --pot {1}/{2}.pot --create {3} {4}'.format(
        i18ndude,
        locale_path,
        domain,
        domain,
        target_path,
    )
    subprocess.call(
        cmd,
        shell=True,
    )


def _sync():
    cmd = '{0} sync --pot {1}/{2}.pot {3}*/LC_MESSAGES/{4}.po'.format(
        i18ndude,
        locale_path,
        domain,
        locale_path,
        domain,
    )
    subprocess.call(
        cmd,
        shell=True,
    )


def update_locale():
    locale_folder_setup()
    _sync()
    _rebuild()


src/collective/edtf_behavior/behaviors/e_d_t_f_date.py
# -*- coding: utf-8 -*-

from collective.edtf_behavior import _
from plone import schema
from plone.autoform.interfaces import IFormFieldProvider
from plone.dexterity.interfaces import IDexterityContent
from plone.supermodel import model
from zope.component import adapter
from zope.interface import Interface
from zope.interface import implementer
from zope.interface import provider


class IEDTFDateMarker(Interface):
    pass

@provider(IFormFieldProvider)
class IEDTFDate(model.Schema):
    """
    """

    project = schema.TextLine(
        title=_(u'Project'),
        description=_(u'Give in a project name'),
        required=False,
    )


@implementer(IEDTFDate)
@adapter(IEDTFDateMarker)
class EDTFDate(object):
    def __init__(self, context):
        self.context = context

    @property
    def project(self):
        if hasattr(self.context, 'project'):
            return self.context.project
        return None

    @project.setter
    def project(self, value):
        self.context.project = value


src/collective/edtf_behavior/behaviors/edtf_date.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from collective.edtf_behavior import _
from plone import schema
from plone.autoform.interfaces import IFormFieldProvider
from plone.supermodel import model
from zope.component import adapter
from zope.interface import implementer
from zope.interface import Interface
from zope.interface import Invalid
from zope.interface import provider

from edtf2.parser.edtf_exceptions import EDTFParseException
from edtf2 import parse_edtf, struct_time_to_date
from datetime import date
import math


class IEDTFDateMarker(Interface):
    pass


def edtf_parseable(value):
    """ make sure value is parseable by parse_edtf
    """

    try:
        parse_edtf(value)
    except EDTFParseException:
        raise Invalid(
            _(u'Invalid EDTF date format!'),
        )
        return False
    else:
        return True


@provider(IFormFieldProvider)
class IEDTFDate(model.Schema):
    """
    """

    edtf_date = schema.TextLine(
        title=_(u'Date'),
        description=_(
            u'Enter a date or date interval in <a target="_blank" href="http://www.loc.gov/standards/datetime/edtf.html">EDTF</a> format. Examples: "1860-03-31", "1860-03~," or "1860-03-31/1860-04-15"',  # NOQA E501
        ),
        required=False,
        constraint=edtf_parseable,
    )


@implementer(IEDTFDate)
@adapter(IEDTFDateMarker)
class EDTFDate(object):
    def __init__(self, context):
        self.context = context

    @property
    def edtf_date(self):
        if hasattr(self.context, 'edtf_date'):  # NOQA: P002
            return self.context.edtf_date
        return None

    @edtf_date.setter
    def edtf_date(self, value):
        edtf_date = value
        self.context.edtf_date = edtf_date

    @property
    def date_earliest(self):
        if not self.context.edtf_date:
            return
        edtf_obj = parse_edtf(self.context.edtf_date)
        result = edtf_obj.lower_fuzzy()
        if result == -math.inf:
            return date.min
        return struct_time_to_date(result)

    @property
    def date_latest(self):
        if not self.context.edtf_date:
            return
        edtf_obj = parse_edtf(self.context.edtf_date)
        result = edtf_obj.upper_fuzzy()
        if result == math.inf:
            return date.max
        return struct_time_to_date(result)

    @property
    def date_sort_ascending(self):
        if not self.context.edtf_date:
            return
        edtf_obj = parse_edtf(self.context.edtf_date)
        result = edtf_obj.lower_strict()
        if result == -math.inf:
            return date.min
        return struct_time_to_date(result)

    @property
    def date_sort_descending(self):
        if not self.context.edtf_date:
            return
        edtf_obj = parse_edtf(self.context.edtf_date)
        result = edtf_obj.upper_strict()
        if result == math.inf:
            return date.max
        return struct_time_to_date(result)
